ui/src/app/store/zone/slice.ts

```diff
@@ -1,22 +1,291 @@
+import type { PayloadAction } from "@reduxjs/toolkit";
 import { createSlice } from "@reduxjs/toolkit";
 
-import { ZoneMeta } from "./types";
-import type { CreateParams, UpdateParams, ZoneState } from "./types";
-
 import {
-  generateCommonReducers,
-  genericInitialState,
-} from "app/store/utils/slice";
+  ZONE_ACTIONS,
+  ZONE_MODEL,
+  ZONE_PK,
+  ZONE_WEBSOCKET_METHODS,
+} from "./constants";
+import type {
+  CreateParams,
+  DeleteParams,
+  UpdateParams,
+  Zone,
+  ZoneActionNames,
+  ZonePayloadActionWithMeta,
+  ZoneGenericActions,
+  ZoneModelActions,
+  ZonePK,
+  ZoneState,
+} from "./types";
+
+import { ACTION_STATUS } from "app/base/constants";
+import type { ActionStatuses, APIError } from "app/base/types";
+
+const {
+  cleanup: cleanupAction,
+  create: createAction,
+  delete: deleteAction,
+  fetch: fetchAction,
+  update: updateAction,
+} = ZONE_ACTIONS;
+
+const {
+  error: errorStatus,
+  loading: loadingStatus,
+  success: successStatus,
+} = ACTION_STATUS;
+
+export const initialGenericActions: ZoneGenericActions = {
+  [createAction]: ACTION_STATUS.idle,
+  [fetchAction]: ACTION_STATUS.idle,
+};
+
+export const initialModelActions: ZoneModelActions = {
+  [deleteAction]: {
+    [errorStatus]: [],
+    [loadingStatus]: [],
+    [successStatus]: [],
+  },
+  [updateAction]: {
+    [errorStatus]: [],
+    [loadingStatus]: [],
+    [successStatus]: [],
+  },
+};
+
+const addError = (
+  state: ZoneState,
+  action: ZoneActionNames,
+  error: APIError,
+  modelPK: ZonePK | null = null
+) => {
+  state.errors.unshift({ action, error, modelPK });
+};
+
+const updateGenericAction = (
+  state: ZoneState,
+  actionName: keyof ZoneGenericActions,
+  status: ActionStatuses
+) => {
+  state.genericActions[actionName] = status;
+};
+
+const updateModelAction = (
+  state: ZoneState,
+  actionName: keyof ZoneModelActions,
+  status: ActionStatuses,
+  modelPK: ZonePK
+) => {
+  const statuses = state.modelActions[actionName];
+  statuses[errorStatus] = statuses[errorStatus].filter((id) => id !== modelPK);
+  statuses[loadingStatus] = statuses[loadingStatus].filter(
+    (id) => id !== modelPK
+  );
+  statuses[successStatus] = statuses[successStatus].filter(
+    (id) => id !== modelPK
+  );
+  if (status === errorStatus) {
+    statuses[errorStatus].push(modelPK);
+  } else if (status === loadingStatus) {
+    statuses[loadingStatus].push(modelPK);
+  } else {
+    statuses[successStatus].push(modelPK);
+  }
+};
 
 const zoneSlice = createSlice({
-  name: ZoneMeta.MODEL,
-  initialState: genericInitialState as ZoneState,
-  reducers: generateCommonReducers<
```

Comment:
Yeah for sure. I'm a little hesitant about recreating the `generateCommonReducers()` util though because it's a bit opaque - I like the idea of having all the action names + reducers in one file, clearly outlined. Maybe just utils for generating single reducers rather than a collection of reducers?

Reply:
The other option might be to provide an array of reducers to include so that is clear which ones apply to that model, something like `generateCommonReducers(["list", "getById", "delete"])`. I remember proposing that back when I was building the utils the first time, but I don't think there was a specific reason why we didn't implement it.

Reply:
Could do that, but even in that case you wouldn't get the "true" list unless you included `["delete", "deleteStart", "deleteSuccess", "deleteError"]` etc.

Reply:
Yeah though I'm not sure how much value there is in explicitly seeing those...